website/blog/2022-11-31-simple-2.mdx

```diff
@@ -0,0 +1,115 @@
+
+To show how much simpler simple functions are. The figure below shows a function _NestedMapSum_ that is written in both the simple and vector interfaces. The function consumes a nested map and computes the summations of all values and keys.
+Note that the vector function implementation is minimal, without any special optimization. Adding optimizations will make it even longer.
+
+<figure>
+ <img src="/img/simple1_2.png"/>
+ <figcaption> NestedMapSum function implemented using vector(left) and simple(right) interfaces.</figcaption>
+</figure>
+
+
+### View types for inputs
+The previous representations of input complex types in the simple function interface were extremely expensive. Data from vectors used to be copied into std containers and passed to simple functions to process it.
+
+For example, to execute the function _array_first_ that returns the first element in an array, the SimpleFunctionAdapter used to read the input array from the input vector for every row and construct an _std::vector<std::optional<\>\>_, (std::optional is used to express the nullity) and pass it to the function for execution. The graph below illustrates the previous implementation approach.
+<figure>
+ <img src="/img/simple1_3.png"/>
+</figure>
+similarly maps and rows used to be materialized into folly::F14FastMap and std::tuples. The previous approach has two key inefficiencies:
+
+**Eager materialization** : For each row, all the data in the input vector is decoded and read before calling the function. However, in some cases functions might not need all the data to be read. For example, the function _array_first_ only needs to read the first element in every array, moreover the _cardinality_ function does not need to read any elements in the array.
+
+**Double reading** : The data is read twice once when the input is constructed, and again in the function when it's used.
+
+Such an approach used to make simple functions significantly slower, up to more than `10X` for input maps. In order to mitigate those regressions, Velox introduced <a href="https://github.com/facebookincubator/velox/blob/main/velox/expression/ComplexViewTypes.h">View types</a>: ArraViews, MapViews ...etc. The goal is to keep the authoring simple but achieve **at least** the performance of a basic vector implementation that decodes input and applies some logic for every row without any special optimizations.
+
+The new view types are **Lazy**, very cheap to construct and do not materialize the underlying data unless the code accesses it. They are also designed to have interfaces similar to those of std::containers.
+
+In a simplified form, an ArrayView stores the length and the offset of the array within the vector, in addition to a pointer to the elements array. Only when an element is accessed then an OptionalAccessor is created, which contains the index of the accessed element and a pointer
```

Comment:
An OptionalAccessor is only created, containing the index of the accessed element and a pointer, when an element is accessed.